cloudapi/config.go

```diff
@@ -346,3 +341,83 @@ func GetConsolidatedConfig(
 
 	return result, nil
 }
+
+// mergeFromCloudOptionAndExternal merges three fields from the JSON in a cloud key of
+// the provided external map. Used for options.cloud settings.
+func mergeFromCloudOptionAndExternal(
+	cloudConfig json.RawMessage,
+	external map[string]json.RawMessage,
+	conf *Config,
+) error {
+	source := pickSource(cloudConfig, external)
+	if source == nil {
+		return nil
+	}
+
+	// Original comment
+	// TODO: Important! Separate configs and fix the whole 2 configs mess!
+	tmpConfig := Config{}
+	if err := json.Unmarshal(source, &tmpConfig); err != nil {
+		return err
+	}
+	// Only take out the ProjectID, Name and Token from the options.cloud (or legacy loadimpact struct) map:
+	if tmpConfig.ProjectID.Valid {
+		conf.ProjectID = tmpConfig.ProjectID
+	}
+	if tmpConfig.Name.Valid {
+		conf.Name = tmpConfig.Name
+	}
+	if tmpConfig.Token.Valid {
+		conf.Token = tmpConfig.Token
+	}
+
+	return nil
+}
+
+// GetTemporaryCloudConfig returns a temporary cloud config.
+// Original comment
+// TODO: Fix this
+// We reuse cloud.Config for parsing options.cloud (or legacy loadimpact struct), but this probably shouldn't be
+// done, as the idea of options.ext is that they are extensible without touching k6. But in
+// order for this to happen, we shouldn't actually marshall cloud.Config on top of it, because
+// it will be missing some fields that aren't actually mentioned in the struct.
+// So in order for use to copy the fields that we need for k6 cloud's api we unmarshal in
+// map[string]interface{} and copy what we need if it isn't set already
+func GetTemporaryCloudConfig(
+	cloudConfig json.RawMessage,
+	external map[string]json.RawMessage,
+) (map[string]interface{}, error) {
+	tmpCloudConfig := make(map[string]interface{}, 3)
+
+	source := pickSource(cloudConfig, external)
+	if source == nil {
+		return tmpCloudConfig, nil
+	}
+
+	dec := json.NewDecoder(bytes.NewReader(source))
+	dec.UseNumber() // otherwise float64 are used
+	if err := dec.Decode(&tmpCloudConfig); err != nil {
+		return nil, err
+	}
+
+	return tmpCloudConfig, nil
+}
+
+// pickSource returns the config source to use.
+func pickSource(
+	cloudConfig json.RawMessage,
+	external map[string]json.RawMessage,
+) json.RawMessage {
+	// priority is the new way of defining the config
+	// via opt.config
```

Comment:
I have the feeling this comment should be rearranged. It seems to me `priority` is an option and could be set somewhere, but isn't, right?

Reply:
Good catch :facepalm: